compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,214 @@
+        for l in body.vars_and_temps_iter() {
+            let l_decl = &mut body.local_decls[l];
+            let (num_variants, sizes, match_ty): (u64, Vec<_>, _) = match l_decl.ty.kind() {
+                ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                    let p_e = tcx.param_env(adt_def.did);
+                    // FIXME(jknodt) handle error better below
+                    let variants = &tcx.layout_of(p_e.and(l_decl.ty)).unwrap().variants;
+                    match variants {
+                        Variants::Single { .. } => continue,
+                        Variants::Multiple { variants, .. } if variants.len() <= 1 => continue,
+                        Variants::Multiple { variants, .. } => {
+                            let min = variants.iter().map(|v| v.size).min().unwrap();
+                            let max = variants.iter().map(|v| v.size).max().unwrap();
+                            if max.bytes() - min.bytes() < D {
+                                continue;
+                            }
+                            (
+                                variants.len() as u64,
+                                variants.iter().map(|v| v.size).collect(),
+                                l_decl.ty,
+                            )
+                        }
+                    }
+                }
+                _ => continue,
+            };
+            // Has matched
+            let mut patch = MirPatch::new(body);
+            let (bbs, local_decls) = body.basic_blocks_and_local_decls_mut();
+            for bb in bbs {
+                bb.expand_statements(|st| {
+                    match &st.kind {
+                        StatementKind::Assign(box (
+                            lhs,
+                            Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
+                        )) if lhs.ty(local_decls, tcx).ty == match_ty => {
+                            let source_info = st.source_info;
+                            let span = source_info.span;
+
+                            let tmp_ty = tcx.mk_ty(ty::Array(
+                                tcx.types.usize,
+                                Const::from_usize(tcx, num_variants),
+                            ));
+
+                            let new_local = patch.new_temp(tmp_ty, span);
+                            let store_live = Statement {
+                                source_info,
+                                kind: StatementKind::StorageLive(new_local),
+                            };
+
+                            let place = Place { local: new_local, projection: List::empty() };
+                            let mut data =
+                                vec![0; std::mem::size_of::<usize>() * num_variants as usize];
+                            data.copy_from_slice(unsafe { std::mem::transmute(&sizes[..]) });
+                            let alloc = interpret::Allocation::from_bytes(
```

Comment:
For the vector of sizes? I might be missing something but how would the alignment matter in that case? I can see it mattering in the case of the enum itself, but otherwise I'm not sure.

Reply:
Regarding alignment, I was referring to the alignment of source & destination pointers when copying the enum itself. This appears to be hard to express on the MIR level with CopyNonOverlapping, since alignment cannot be provided independently of the pointer type.

Reply:
Does the alignment matter when just performing a memcpy of bytes though? I guess I usually don't see alignment issues because they're handled automagically but